File: Codeforces/1996/decode.go
// https://codeforces.com/contest/1996/problem/E

package main
import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

const MOD int64 = 1e9 + 7

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()
	scanner.Scan()
	T, _ := strconv.Atoi(scanner.Text())

	for t := 0; t < T; t++ {
		scanner.Scan()
		S := scanner.Text()
		N := len(S)
		freq := make(map[int64]int64)
		cur := int64(0)
		ans := int64(0)
		freq[0] = 1

		for i := 0; i < N; i++ {
			if S[i] == '1' {
				cur++
			} else {
				cur--
			}
			if val, found := freq[cur]; found {
				ans = (ans + val * int64(N-i) % MOD) % MOD
			}
			freq[cur] += int64(i + 2)
		}
		fmt.Fprintln(writer, ans)
	}
}

File: Codeforces/1996/decode.cpp
// https://codeforces.com/problemset/problem/1996/E

#include <bits/stdc++.h>
using namespace std;
#define int long long
const int MOD = 1e9 + 7;

signed main() {
    cin.tie(0)->sync_with_stdio(0);
    int T; cin >> T;
    while (T--) {
        string S;
        cin >> S;
        int N = S.length();
        unordered_map<int64_t, int64_t> freq;
        int cur = 0;
        int ans = 0;
        freq[0] = 1;

        for (int i = 0; i < N; i++) {
            cur += (S[i] == '1' ? 1 : -1);
            if (freq.find(cur) != freq.end()) {
                ans = (ans + freq[cur] * (N - i) % MOD) % MOD;
            }
            freq[cur] += i + 2;
        }
        cout << ans << '\n';
    }
}
